hydra-node/test/Hydra/JSONSchema.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-unused-imports #-}
+{-# OPTIONS_GHC -Wno-unused-local-binds #-}
+
+module Hydra.JSONSchema where
+
+import Hydra.Prelude
+
+import Control.Lens (Lens', to, (^?))
+import qualified Data.Aeson as Aeson
+import Data.Aeson.Lens (AsValue, key, _Array, _String)
+import Data.Char (isSpace)
+import Data.List (dropWhileEnd)
+import qualified Data.Map.Strict as Map
+import qualified Data.Yaml as Yaml
+import qualified Paths_hydra_node as Pkg
+import System.Exit (ExitCode (..))
+import System.FilePath ((</>))
+import System.Process (readProcessWithExitCode)
+import Test.Hspec (pendingWith)
+import Test.Hydra.Prelude (createSystemTempDirectory)
+import Test.QuickCheck (Property, conjoin, counterexample, forAllBlind, forAllShrink, vectorOf, withMaxSuccess)
+import Test.QuickCheck.Monadic (assert, monadicIO, monitor, run)
+import qualified Prelude
+
+-- | Generate arbitrary serializable (JSON) value, and check their validity
+-- against a known JSON schema.
+-- This property ensures that JSON instances we produce abide by
+-- the specification. Note this, because this uses an external tool each
+-- property iteration is pretty slow. So instead, we run the property only
+-- once, but on a list of 100 elements all arbitrarily generated.
+prop_validateToJSON ::
+  forall a.
+  (ToJSON a, Arbitrary a, Show a) =>
+  FilePath ->
+  FilePath ->
+  Property
+prop_validateToJSON specFile inputFile =
+  forAllShrink (vectorOf 100 arbitrary) shrink $ \(a :: [a]) ->
+    monadicIO $
+      do
+        run ensureSystemRequirements
+        let obj = Aeson.encode a
+        (exitCode, _out, err) <- run $ do
+          writeFileLBS inputFile obj
+          readProcessWithExitCode "jsonschema" ["-i", inputFile, specFile] mempty
+        monitor $ counterexample err
+        monitor $ counterexample (decodeUtf8 obj)
+        assert (exitCode == ExitSuccess)
+
+-- | Check specification is complete wr.t. to generated data
+-- This second sub-property ensures that any key found in the
+-- specification corresponds to a constructor in the corresponding
+-- data-type. This in order the document in sync and make sure we don't
+-- left behind constructors which no longer exists.
+--
+-- The second argument is a lens-like function that says which part of the
+-- specification to select to check completeness of the specification w.r.t.
+-- constructors for the datatype, for example:
+--
+-- @@
+-- key "properties" . key "message" . key "oneOf" . _Array . to toList
+-- @@
+--
+-- which selects the list of elements under @properties > message > oneOf@ path
+-- in the specification file.
```

Comment:
Could: Maybe documenting how the spec is to be structured to use this property is more easier to follow? I am not a novice developer, but do get confused by this documentation.